uom edit bugfix for measuredValues and linkedUOMs

Shahad7 opened the pull request:
Changes :-

- Created separate method for editing measuredValues 
- Added @Id and @TypeName annotations to UnitClassLinkageDtoV2 and UnitClassSelfLinkageDtoV2 respectively so
  that JaVers can create a normal globalId per linkedUOM objects
 - @JsonProperty annotations were added to UnitClassMeasuredValuesDtoV2 as exceptions were happening in 
   hibernate queries
 - In UOMEntity, on deleting linkedUOM entities fromUOMs had to be cleared instead of setting it to null as it was 
   throwing exception : "A collection with cascade=”all-delete-orphan” was no longer referenced by the owning entity 
   instance"

Shahad7 committed:
uom edit bugfix for measuredValues and linkedUOMs

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/entity/UOMEntity.java:
```diff
@@ -163,7 +163,7 @@ public void removeLinkedFromUOMs() {
                 f.setToUom(null);
             });
-            this.fromUOMs = null;
+            this.fromUOMs.clear();
```
check delete uom API is working or not with this change

Shahad7 replied:
it's working

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+                }
+            }));
+
+            diffCollections.getChangesByType(NewObject.class).forEach(Errors.rethrow().wrap(p->{
```
what is NewObject?

Shahad7 replied:
Change when a  new object is added to the List

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+                Optional<UOMEntity> optionalTo = uomJpaRepository.findByCode(elt.getCode());
+                if(optionalTo.isPresent()){
+                    UOMEntity to = optionalTo.get();
+                    UOMSelfLinkageEntity selfLinkage = UOMSelfLinkageEntity.builder()
```
use bean utils to copy property from source to target instead of using property specific getters and setts as this logic needs to be dynamic for extending to other types of resources in other sub domains

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/validator/MutuallyInclusiveMeasuredValuesValidatorImpl.java:
```diff
@@ -41,7 +41,7 @@ public boolean isValid(List<? extends MeasuredValuesContract> value, ConstraintV
         boolean hasCorrectIndices = measurementSystems.stream().map(f -> {
             Integer indexOfMeasuredValuesForMeasurementSystem = IntStream.range(0, value.size()).filter(i -> value.get(i).getMeasurementSystem().compareTo(f) == 0).findFirst().getAsInt();
-            return indexOfMeasuredValuesForMeasurementSystem != f.getOrdinal();
+            return indexOfMeasuredValuesForMeasurementSystem == f.getOrdinal();
```
why is the condition reversed?

Shahad7 replied:
it was throwing validation error

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
             log.debug(diffCollections.prettyPrint());
+            diffCollections.getChangesByType(ValueChange.class).forEach(Errors.rethrow().wrap(p -> {
+
+                if(!(p instanceof InitialValueChange || p instanceof TerminalValueChange) ){
```
use polymorphism with helper method to replace instanceof checks

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+        return target;
+    }
+
+    private UOMEntity updateMeasuredValues(UOMDtoV2 old,UOMDtoV2 _new, UOMEntity target) {
```
what was wrong with the old logic for updating measured values?

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+            log.debug("{} changed from {} to {}", p.getPropertyNameWithPath(), p.getLeft(), p.getRight());
+            String propPath = p.getPropertyNameWithPath();
+            int currentIndex = Integer.parseInt(propPath.substring(propPath.indexOf("/")+1,propPath.indexOf("/")+2));
+            beanUtilsBean.copyProperty(target.getMeasuredValues().get(currentIndex),p.getPropertyName(),p.getRight());
```
can't you use bean utils to replace elements in collection directly instead of replacing them individually by their indices?

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+            diffCollections.getChangesByType(ValueChange.class).forEach(Errors.rethrow().wrap(p -> {
+
+                if(!(p instanceof InitialValueChange || p instanceof TerminalValueChange) ){
+                    if(p.getPropertyName().equals("quantity")) {
```
do not check for individual property names, as this logic needs to be dynamic for extending to other types of resources in other sub domains

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/service/impl/UOMServiceImpl.java:
```diff
@@ -361,31 +363,90 @@ public void updateExistingUOMByCode(String code, UOMDtoV2 sourceUOMDto) {
+                if(!(p instanceof InitialValueChange || p instanceof TerminalValueChange) ){
+                    if(p.getPropertyName().equals("quantity")) {
+                        log.debug("{} changed from {} to {}", p.getPropertyNameWithPath(), p.getLeft(), p.getRight());
+                        int newQuantity = Integer.parseInt(p.getRight().toString());
```
use beanutils to copy property from source to target instead of using property specific getters and setts as this logic needs to be dynamic  for extending to other types of resources in other sub domains

Shahad7 committed:
status field not null validation

Shahad7 committed:
measuredValues manual update method call removed and ignored in javers diff temporarily

Shahad7 committed:
longName field added to UnitClassDtoV2

Shahad7 committed:
progress

Shahad7 committed:
replaced with json merge patch

Shahad7 committed:
tmp merge

Shahad7 committed:
added empty default values to all fields in UOMDtoV2 so that json merge patch won't throw no such path exception

Shahad7 committed:
linkedUOM field in converter

Shahad7 committed:
clean checkpoint

Shahad7 committed:
fixed update method implementation to work comparitivelyUpdateMandatoryFields handler when json merge patch is used

Shahad7 committed:
checkpoint with comparitivelyCollectionUpdateHandler v1

Shahad7 committed:
extracted update logic to a separate method

Shahad7 commented:
Changes :- 

- ComparisonConfigDto class for storing details needed for comparing collection fields
- ComparitveCollectionUpdateHandler with default method for comparing collections
- Introduced a new contract IdentifiableCollectionItem so that all entity collection fields returns a unique id and same 
  with respective dto collection fields
- In UnitClassLinkageDtoV2 class, added fromUOM's code as well, since it was needed for the reducer's logic 
  (to look up fromUOMEntity)
- Introduced CollectionItemDtoToEntityReducer as a contract; used in ComparitveCollectionUpdateHandler.
- callSuper attributed was removed from EqualsAndHashCode and ToString annotations of certain nested entities as  
  they were leading to stackOverFlow errors

Shahad7 committed:
Deleted UnitClassCollectionItem contract

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/constants/HttpMediaType.java:
```diff
@@ -4,5 +4,6 @@ public class HttpMediaType {
 
     public static final String APPLICATION_JSON_PATCH = "application/json-patch+json";
```
is this required?

Shahad7 replied:
not anymore

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/converter/UOMEntityToDtoV2Converter.java:
```diff
@@ -47,8 +52,11 @@ public UOMDtoV2 convert(UOMEntity source) {
                 .isPurchase(source.getIsPurchase())
                 .status(source.getStatus())
                 .measuredValues(measuredValues)
```
Remove measuredValues field from DTO and adjust logic accordingly. Measured values should exist as a dedicated resource with loose coupling to UOMDto by not being a collection property of UOMDto of type MeasuredValues but rather highly cohesive with the UOMDTo's id as. a property of MeasuredValues

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/converter/UOMEntityToDtoV2Converter.java:
```diff
@@ -47,8 +52,11 @@ public UOMDtoV2 convert(UOMEntity source) {
                 .isPurchase(source.getIsPurchase())
                 .status(source.getStatus())
                 .measuredValues(measuredValues)
+                .linkedUOMs(linkedUOMs)
```
Remove linkedUOMs field from DTO and adjust logic accordingly. Linked UOM should exist as a dedicated resource with loose coupling to UOMDto by not being a collection property of UOMDto of type UOMCrossLinkage and UOMSelfLinkage but rather highly cohesive with the UOMDTo's id as a property of UOMCrossLinkage and UOMSelfLinkage

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/dto/UOMDtoV2.java:
```diff
@@ -42,15 +44,16 @@ public class UOMDtoV2 extends UnitClassDtoV2 {
             description = "this UOM is linked to these UOMs"))
     @DiffIgnore
-    private Optional<@Size(min = 1, message = "At least 1 UOM must be linked") List<@Valid UnitClassSelfLinkageDtoV2>> linkedUOMs = Optional.empty();
+    private @Size(min = 1, message = "At least 1 UOM must be linked") List<@Valid UnitClassSelfLinkageDtoV2> linkedUOMs;
```
UOMDTo should no longer contain linkedUOMs field to facilitate loose coupling for easy management of UnitClassSelfLinkageDtoV2

anirbandas18 commented on src/main/java/com/teenthofabud/wizard/nandifoods/wms/settings/unit/uom/dto/UOMDtoV2.java:
```diff
@@ -42,15 +44,16 @@ public class UOMDtoV2 extends UnitClassDtoV2 {
-    private Optional<@Size(min = 1, message = "At least 1 PU/HU must be linked") Set<@Valid UnitClassCrossLinkageDtoV2>> linkedPUHUs = Optional.empty();
+    @DiffIgnore
+//    @Builder.Default
+    private @Size(min = 1, message = "At least 1 PU/HU must be linked") List<@Valid UnitClassCrossLinkageDtoV2> linkedPUHUs;
```
UOMDto should no longer contain linkedPUHUs field to facilitate loose coupling for easy management of UnitClassCrossLinkageDtoV2
